org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/DefaultGradleBuild.java

```diff
@@ -83,9 +97,40 @@ public DefaultGradleBuild(org.eclipse.buildship.core.internal.configuration.Buil
 
     @Override
     public SynchronizationResult synchronize(IProgressMonitor monitor) {
+        IStatus toolingApiCompatibilityStatus = validateToolingApiCompatibility(monitor);
+        if (!toolingApiCompatibilityStatus.isOK()) {
+            return DefaultSynchronizationResult.from(toolingApiCompatibilityStatus);
+        }
         return synchronize(NewProjectHandler.IMPORT_AND_MERGE, GradleConnector.newCancellationTokenSource(), monitor);
     }
 
+    private IStatus validateToolingApiCompatibility(IProgressMonitor monitor) {
+        String testingProperty = System.getProperty("eclipse.pde.launch");
+        if (testingProperty == null) {
+            String gradleVersion;
+            try {
+                BuildEnvironment environment = this.withConnection(connection -> connection.getModel(BuildEnvironment.class), monitor);
+                gradleVersion = environment.getGradle().getGradleVersion();
+            } catch (Exception e) {
+                return ToolingApiStatus.from("Project synchronization", e);
+            }
+            if (gradleVersion == null) {
+                return ToolingApiStatus.from("Project synchronization", new GradleConnectionException("Can't determine Gradle version when synchronizing project."));
+            }
+            File javaHome = this.buildConfig.getJavaHome();
+            String javaVersion = javaHome != null ? new StandardVMType().readReleaseVersion(javaHome) : System.getProperty("java.version");
+            if (javaVersion == null) {
+                return ToolingApiStatus.from("Project synchronization", new UnsupportedJavaVersionException("Can't determine Java version when synchronizing project."));
+            }
+            JavaVersion javaVersionObject = JavaVersion.toVersion(javaVersion);
+            String minGradleVersion = compatibilityMap.get(javaVersionObject.getMajorVersion());
+            if (minGradleVersion != null && GradleVersion.version(gradleVersion).compareTo(GradleVersion.version(minGradleVersion)) < 0) {
+                return ToolingApiStatus.from("Project synchronization", new UnsupportedJavaVersionException(String.format("Can't use Java %s and Gradle %s to synchronize project.", javaVersion, gradleVersion)));
```

Comment:
💄 Let's make the exception message more actionable.

```suggestion
                return ToolingApiStatus.from("Project synchronization", new UnsupportedJavaVersionException(String.format("The current build uses Gradle %s running on Java %s which is not supported. Please consult the Gradle documentation to find the compatible combinations: https://docs.gradle.org/current/userguide/compatibility.html.", gradleVersion, javaVersion)));
```